### ISSUE TITLE
Undefined control sequence, but don't display more detail

### ISSUE THREAD

**Issue by Freed-Wu** [lervag/vimtex#2526]
### Description

When I `latexmk`,

```
./2022-10-21.tex:89: Undefined control sequence.
<recently read> \cdashline

l.89 \end{frame}

?
```

However, in vimtex, it don't display any information about this control 
sequence's name.

```
...
2022-10-21.tex|89 error| Undefined control sequence.
2022-10-21.tex|89 warning| Overfull \hbox (68.54958pt too wide) in paragraph at lines 89--89
...
```

I think `&errorfmt` should be rewrite to display its name.

Thanks!

### Do you use a latexmkrc file?

Yes

### VimtexInfo

```yaml
System info:
  OS: Arch Linux
  Vim version: NVIM v0.8.0
  Has clientserver: true
  Servername: /run/user/1000/nvim.1399382.0

VimTeX project: 2022-10-17
  base: 2022-10-17.tex
  root: /home/wzy/Documents/weekly-report
  tex: /home/wzy/Documents/weekly-report/2022-10-17.tex
  main parser: recursive search (bib)
  document class: article
  packages: array autofancyhdr geometry hyperref longtable natbib parskip xcolor xspace
  compiler: latexmk
    engine: -xelatex
    options:
      -verbose
      -file-line-error
      -synctex=1
      -interaction=nonstopmode
    callback: 1
    continuous: 1
    executable: latexmk
  viewer: Zathura
    xwin id: 0
  qf method: LaTeX logfile
```

**Comment by Freed-Wu** [comment]
```tex
\documentclass{beamer}
\usetheme{Frankfurt}
\usecolortheme{spruce}

\begin{document}

\begin{frame}
  \frametitle{Structure}
  \begin{table}[ht!]
    \centering
    \begin{tabular}{lcccccc}
      \cdashline{1-7}
    \end{tabular}
  \end{table}
\end{frame}
\end{document}
```

```vim
#!/usr/bin/env -S vi -u
" $ uname -r
" 5.19.7-arch1-1
" $ has
" ✓ vi 0.8.0
" $ cat test.vim
set runtimepath=$VIMRUNTIME
set runtimepath+=~/.local/share/nvim/repos/github.com/lervag/vimtex
" $ chmod +x test.vim
" $ ./test.vim test.tex
```

```vim
:call vimtex#init()
:VimtexCompile
```

```
test.tex|15 error| Undefined control sequence.
```

```sh
$ latexmk -gg *.tex
...
./a.tex:15: Undefined control sequence.
\beamer@doifinframe ...bular}{lcccccc} \cdashline
                                                  {1-7} \end {tabular} \end ...
l.15 \end{frame}

?
```

**Comment by lervag** [comment]
Ah. In this case, parsing is very hard. The lines to parse are these:

```
1 ./a.tex:15: Undefined control sequence.
2 \beamer@doifinframe ...bular}{lcccccc} \cdashline
3                                                   {1-7} \end {tabular} \end ...
4 l.15 \end{frame}
```

The way `errorfmt` works is that it parses each line _separately_. Thus, first line 1 is parsed, and it is recognized as an error. Then line 2 is parsed. If it is recognized as a continuation of an error, then it can be combined in a multiline error. However, both lines 2 and 3 are more or less random, and there is no way (that I know) to just ignore them. Line 4, on the other hand, is a known continuation. and _is_ parsed already. However, since lines 2 and 3 were not parsed, line 4 will not match any rules (because it is a continuation).

What you ask requires to add an `errorfmt` entry that can parse line 2 (line 3 is OK to ignore). To parse, we can only use a relatively simple set of building blocks (see `:help error-file-format`). And to be honest, I don't see how to make a rule based on this language that works here.

Sorry for the inconvenience!

**Comment by Freed-Wu** [comment]
Maybe we need some external tools (like `pplatex`) to parse the log and return a easy-handled result to vim errorformat.

**Comment by Freed-Wu** [comment]
I try pplatex, it still don't separate `\cdashline`

```
❯ pplatex -c latexmk test.tex
warning: command line (kpathsea): No cnf value in argument: latexmk.
** Error   in ./test.tex, Line 15:
   Undefined control sequence \beamer@doifinframe ...bular}{lcccccc} \cdashline
   {1-7} \end {tabular} \end ...
   \end{frame}

Result: o) Errors: 1, Warnings: 0, BadBoxes: 0
        o) latex returned an error!
```

How about just treat the second line as the result of undefined control sequence?

Just like

```
test.tex|15 error| Undefined control sequence. \beamer@doifinframe ...bular}{lcccccc} \cdashline
```

It is worse than

```
test.tex|15 error| Undefined control sequence. \cdashline
```

but better than nothing.

**Comment by Rmano** [comment]
Well, (I think I commented something like that in the far past), the "split line" is the standard way that TeX (the lower level) use to signal *which* is the undefined macro. 

Consider that TeX was born when there was no highlight, no graphic cursors, no color, and often a 1200 baud unidirectional (in the sense that it can't back up) terminal was the only interactive interface. So Knuth decided that the way to signal the error was to "split print" the erroring line, stopping the first one after the erroring control sequence and continuing in the next line (if you think at a printing terminal, it was just a matter of issuing a Line Feed without a Carriage Return). In the '80ies, every byte counted.

This is also why the error log in TeX is hard-line-splitted (urgh). That one can be avoided or at least mitigated, but the split line on an undefined error, I think it's so ingrained that it's impossible to change unless you write some kind of preprocessor. It's a bit that so difficult to parse in Vim, but I also resigned on having to `\lo` to know the name of the macro.

Now, if it were possible to open the `\lo` windows on the correct line... that would be awesome.

**Comment by lervag** [comment]
First, thanks @Rmano for the useful insights and explanations on the history of the current behaviour!

> Maybe we need some external tools (like `pplatex`) to parse the log and return a easy-handled result to vim errorformat.

VimTeX already has support for `pplatex` as an external filter, see `:help g:vimtex_quickfix_method`.

> I try pplatex, it still don't separate `\cdashline`
> 
> ```
> ❯ pplatex -c latexmk test.tex
> warning: command line (kpathsea): No cnf value in argument: latexmk.
> ** Error   in ./test.tex, Line 15:
>    Undefined control sequence \beamer@doifinframe ...bular}{lcccccc} \cdashline
>    {1-7} \end {tabular} \end ...
>    \end{frame}
> 
> Result: o) Errors: 1, Warnings: 0, BadBoxes: 0
>         o) latex returned an error!
> ```

You could also try `pulp`; not sure if it will be better, but perhaps. Check the mentioned docs (please also notice the Note at the end!).

> How about just treat the second line as the result of undefined control sequence?

No, sorry. Again, the errorformat mechanism for Vim/neovim has some hard limitations. And, as I partly tried to explain, each line is parsed _separately_. This means that a line is parsed without any _context_. So, when we parse the second line, we don't _know_ that the previous line was an undefined error.

**Comment by Freed-Wu** [comment]
I also take a try and found if I must parse `4 l.15 \end{frame}`'s 15,

```
Undefined control sequence.
\beamer@doifinframe ...bular}{lcccccc} \cdashline
```

is hard to parse.

Should we request a feature to (neo)vim to let them support it? Because
I think it is a normal case which should be supported.

**Comment by lervag** [comment]
Yes, feel free to open an issue with neovim. It would be cool to have a better parser for logfiles. I also agree it is not an uncommon case, although it is quite surprising how well the current parser actually works :)
